gui/browserv7/inc/ROOT/RBrowsable.hxx

```diff
@@ -0,0 +1,405 @@
+   RHolder* DoCopy() const final { return new RShared<T>(fShared); }
+public:
+   RShared(T *obj) { fShared.reset(obj); }
+   RShared(std::shared_ptr<T> obj) { fShared = obj; }
+   RShared(std::shared_ptr<T> &&obj) { fShared = std::move(obj); }
+   virtual ~RShared() = default;
+
+   const TClass *GetClass() const final { return TClass::GetClass<T>(); }
+   const void *GetObject() const final { return fShared.get(); }
+};
+
+/** \class RUnique<T>
+\ingroup rbrowser
+\brief Holder of with unique_ptr<T> instance. Should be used to transfer unique_ptr<T> in RBrowsable methods
+\author Sergey Linev <S.Linev@gsi.de>
+\date 2019-10-19
+\warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback is welcome!
+*/
+
+template<class T>
+class RUnique : public RHolder {
+   std::unique_ptr<T> fUnique; ///<! holder without IO
+protected:
+   void *TakeObject() final { return fUnique.release(); }
+public:
+   RUnique(T *obj) { fUnique.reset(obj); }
+   RUnique(std::unique_ptr<T> &&obj) { fUnique = std::move(obj); }
+   virtual ~RUnique() = default;
+
+   const TClass *GetClass() const final { return TClass::GetClass<T>(); }
+   const void *GetObject() const final { return fUnique.get(); }
+};
+
+
+/** \class RAnyObjectHolder
+\ingroup rbrowser
+\brief Holder of any object instance. Normally used with TFile, where any object can be read. Normally RShread or RUnique should be used
+\author Sergey Linev <S.Linev@gsi.de>
+\date 2019-10-19
+\warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback is welcome!
+*/
+
+class RAnyObjectHolder : public RHolder {
+   TClass *fClass{nullptr};   ///<! object class
+   void *fObj{nullptr};       ///<! plain holder without IO
+   bool fOwner{false};        ///<! is object owner
+protected:
+   void *AccessObject() final { return fOwner ? nullptr : fObj; }
+
+   void *TakeObject() final
+   {
+      if (!fOwner)
+         return nullptr;
+      auto res = fObj;
+      fObj = nullptr;
+      fOwner = false;
+      return res;
+   }
+
+   RHolder* DoCopy() const final
+   {
+      if (fOwner || !fObj || !fClass) return nullptr;
+      return new RAnyObjectHolder(fClass, fObj, false);
+   }
+
+public:
+   RAnyObjectHolder(TClass *cl, void *obj, bool owner = false) { fClass = cl; fObj = obj; fOwner = owner; }
+   virtual ~RAnyObjectHolder()
+   {
+      if (fOwner)
+         fClass->Destructor(fObj);
+   }
+
+   const TClass *GetClass() const final { return fClass; }
+   const void *GetObject() const final { return fObj; }
+};
+
+
+class RLevelIter;
+
+
+/** \class RElement
+\ingroup rbrowser
+\brief Basic element of RBrowsable hierarchy. Provides access to data, creates iterator if any
+\author Sergey Linev <S.Linev@gsi.de>
+\date 2019-10-14
+\warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback is welcome!
+*/
+
+class RElement {
```

Comment:
Full name is `ROOT::Experimental::Browsable::RElement`.
Or all classes will have in prefix RBrowsableElement.
Actually, I prefer to move all these header files into "ROOT/Browsable/" include path to avoid any confusion when special "providers" for TFolder should appear

Reply:
OK, fine with me.